samples/star/src/main/kotlin/com/slack/circuit/star/overlay/BottomSheetOverlay.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2022 Slack Technologies, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.slack.circuit.star.overlay
+
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.layout.padding
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.material.ModalBottomSheetLayout
+import androidx.compose.material.ModalBottomSheetValue
+import androidx.compose.material.rememberModalBottomSheetState
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.unit.dp
+import com.slack.circuit.Overlay
+import com.slack.circuit.OverlayNavigator
+import kotlinx.coroutines.launch
+
+class BottomSheetOverlay<Model : Any, Result : Any>(
+  private val model: Model,
+  private val dismissOnTapOutside: Boolean = true,
+  private val onDismiss: (() -> Result)? = null,
+  private val content: @Composable (Model, OverlayNavigator<Result>) -> Unit,
+) : Overlay<Result> {
+  @Composable
+  override fun Content(navigator: OverlayNavigator<Result>) {
+    val sheetState =
+      rememberModalBottomSheetState(
+        initialValue = ModalBottomSheetValue.Hidden,
+        confirmStateChange = {
+          if (dismissOnTapOutside) {
+            if (it == ModalBottomSheetValue.Hidden) {
+              // This is apparently as close as we can get to an "onDismiss" callback, which
+              // unfortunately has no animation
+              val result = onDismiss?.invoke() ?: error("no result!")
+              navigator.finish(result)
```

Comment:
Do we _need_ to do this? It seems to me that `Content` could require an event sink and return the result from the bottom sheet to the requesting presenter, and not need any of the overlay stuff you've done. Am I missing something here?

Reply:
I'm not sure I follow. The overlay doesn't really know anything about the bottomsheet, it's just a thin API to host content at some root level in the compose tree. We need that at some layer to have a hook into this root level, I don't see how we could go without it? I thought about a sink vs simple OverlayNavigator interface and went with the latter because we need a structured way for the overlay API to hook into result events for its own state handling. This is how the compose snackbar works as well, and for the same reason.

Reply:
Talked offline, was missing some context and this is fine 👍